WooCommerce/Classes/ViewRelated/Dashboard/ProductStock/ProductStockDashboardCardViewModel.swift

```
@@ -69,9 +77,114 @@ private extension ProductStockDashboardCardViewModel {
+
+        try await withThrowingTaskGroup(of: Void.self) { [weak self] group in
+            guard let self else { return }
+
+            if variationsToFetchReports.isNotEmpty {
+                group.addTask {
+                    let reports = try await self.fetchVariationReports(
+                        productIDs: Array(Set(variationsToFetchReports.map { $0.parentID })),
+                        variationIDs: variationsToFetchReports.map { $0.productID }
+                    )
+                    for report in reports {
+                        if let variationID = report.variationID {
+                            // In some cases, variation reports can have invalid product ID.
+                            // Fix that by restore the parent ID from the stock item.
+                            let updatedReport: ProductReport = {
+                                guard report.productID == 0,
+                                      let parentID = variationsToFetchReports.first(where: { $0.productID == variationID })?.productID else {
+                                    return report
+                                }
+                                return report.copy(productID: parentID)
+                            }()
+                            self.savedReports[variationID] = updatedReport
+                        }
+                    }
+                }
+            }
+
+            if productsToFetchReports.isNotEmpty {
+                group.addTask {
+                    let reports = try await self.fetchProductReports(productIDs: productsToFetchReports.map { $0.productID })
+                    for report in reports {
+                        self.savedReports[report.productID] = report
+                    }
+                }
+            }
+
+            while !group.isEmpty {
+                // rethrow any failure.
+                try await group.next()
+            }
```

❓ For my learning: Could you help me understand the purpose of this code?

---

I learned this while working on the Performance cards where I had to trigger several network requests in a task group. Without using `try await group.next()`, each task would fail silently. Using this makes sure that the group can rethrow any error occurring in its child tasks. You can find more details in the documentation [here](https://developer.apple.com/documentation/swift/taskgroup/next()).